net,dgram: add ipv6Only option for net and dgram

oyyd opened the pull request:
For our servers, the dual-stack support is enable by default, i.e. binding host "::" will also make "0.0.0.0" bound. This commit add ipv6Only option in `net.Server.listen()`, `net.Socket.connect()` and `dgram.createSocket()` methods which allows to disable dual-stack support. 

Support for cluster module is also provided in this commit.

Fixes: https://github.com/nodejs/node/issues/17664

/cc @silverwind 

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

richardlau commented on test/parallel/test-net-listen-ipv6only.js:
```diff
@@ -0,0 +1,30 @@
+if (!common.hasIPv6)
+  common.skip('no IPv6 support');
+
+// This test ensures that dual-stack suuport is disabled when
```
```suggestion
// This test ensures that dual-stack support is disabled when
```

oyyd commented on doc/api/net.md:
```diff
@@ -266,6 +266,9 @@ added: v0.11.14
     for all users. **Default:** `false`
   * `writableAll` {boolean} For IPC servers makes the pipe writable
     for all users. **Default:** `false`
+  * `ipv6only` {boolean} For TCP servers, setting `ipv6only` to `true` will
```
Though I prefer `ipv6only` here, it's better to follow camel case :)

oyyd replied:
Though I prefer ipv6only here, it's better to follow camel case :)

oyyd replied:
Though I prefer ipv6only here, it's better to follow camel case :)

oyyd replied:
Though I prefer ipv6only here, it's better to follow camel case :)

bnoordhuis commented on lib/internal/cluster/round_robin_handle.js:
```diff
@@ -16,9 +17,14 @@ function RoundRobinHandle(key, address, port, addressType, fd) {
+      port,
+      host: address,
+      // Currently, net module only supports `ipv6Only` option in `flags`.
+      ipv6Only: Boolean(flags | constants.UV_TCP_IPV6ONLY),
```
This looks wrong. `Boolean(flags | constants.UV_TCP_IPV6ONLY)` is always true. Did you mean to use `&`?

oyyd replied:
Yes, a terrible mistake. I'm going to update it and add a test of falsy `ipv6Only` scenario.

bnoordhuis commented on lib/net.js:
```diff
@@ -98,6 +98,10 @@ const {
 
 function noop() {}
 
+function getFlags(ipv6Only = false) {
```
Unused default argument.

oyyd replied:
The default value of `ipv6Only` would be actual `undefined` rather than `false` which is how its document described. This is the only reason I set the default argument.

oyyd commented:
PTAL. (Travis ci task for this halted..)

silverwind commented:
Thanks for doing this. Would you mind als contributing the same option to UDP / dgram?

bnoordhuis commented on doc/api/net.md:
```diff
@@ -266,6 +266,9 @@ added: v0.11.14
   * `writableAll` {boolean} For IPC servers makes the pipe writable
     for all users. **Default:** `false`
+  * `ipv6Only` {boolean} For TCP servers, setting `ipv6Only` to `true` will
+    disable the dual-stack support, i.e., binding host `::` won't make
```
Just "disable dual-stack support", no "the." (And you'd normally say "binding to host", not "binding host".)

bnoordhuis commented on test/parallel/test-cluster-net-listen-ipv6only-false.js:
```diff
@@ -0,0 +1,55 @@
+// This test ensures that the dual-stack support still works for cluster module
+// when `ipv6Only` is not `true`.
+const host = '::';
+const WORKER_ACCOUNT = 3;
```
Did you mean to write `WORKER_COUNT`? The name's a bit confusing to me.

oyyd replied:
I should have been more careful.

jasnell commented on test/parallel/test-cluster-net-listen-ipv6only-false.js:
```diff
@@ -0,0 +1,55 @@
+  const workers = new Map();
+  let address;
+
+  const countdown = new Countdown(WORKER_ACCOUNT, common.mustCall(() => {
```
It's unnecessary to wrap the `Countdown` callback in a `mustCall()`. That's done automatically.

oyyd commented:
I think I misunderstood the comment https://github.com/nodejs/node/issues/17664#issuecomment-406546764. We could bind to and use **unspecified address** and it works as expected:
<details>
<summary>example</summary>

```js
const dgram = require('dgram');

const socket = dgram.createSocket('udp6');
socket.on('message', (msg, rinfo) => {
  console.log('rinfo', rinfo)
});
socket.bind(0, '::', () => {
  const { port } = socket.address();
  const client = dgram.createSocket('udp4');
  // Ensure the `client` won't bind ipv6.
  client.bind(0, () => {
    client.send(Buffer.allocUnsafe(1), port, '0.0.0.0');
  });
});
```

</details>
So I would like to add the option to dgram too. Maybe I should open another PR?

oyyd commented:
`ipv6Only` for `dgram.createSocket()` is added. PTAL.

silverwind commented:
>  Yes, that's possible. But the last discussion #17664 (comment) shows there is no need to provide the same option to dgram. How do you think about this now?

I'd say still go for it. Even if the OS might not properly support the option, we can still expose it.

jasnell commented on test/parallel/test-cluster-net-listen-ipv6only-none.js:
```diff
@@ -0,0 +1,58 @@
+  const workers = new Map();
+  let address;
+
+  const countdown = new Countdown(WORKER_ACCOUNT, common.mustCall(() => {
```
nit: don't need the `common.mustCall()` wrapper here.

jasnell commented on test/parallel/test-cluster-net-listen-ipv6only-rr.js:
```diff
@@ -0,0 +1,58 @@
+  const workers = new Map();
+  let address;
+
+  const countdown = new Countdown(WORKER_ACCOUNT, common.mustCall(() => {
```
nit: don't need the `common.mustCall()` wrapper here.

oyyd commented:
/ping @silverwind @vsemozhetbyt @mscdex @richardlau @jasnell It should look good now. PTAL.

vsemozhetbyt commented:
Docs LGTM.

silverwind commented on doc/api/dgram.md:
```diff
@@ -601,6 +601,9 @@ changes:
                  supported now.
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/23798
+    description: The `ipv6Only` option is supported.    
```
Trailing whitespace here and on the other copies of this.

oyyd replied:
Done!

oyyd commented:
(I guess I should have asked to run the CI before, didn't notice that..)

`parallel/test-net-connect-ipv6only` failed: 
```
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE: address already in use 0.0.0.0:60058
    at Server.setupListenHandle [as _listen2] (net.js:1295:14)
    at listenInCluster (net.js:1343:12)
    at doListen (net.js:1482:7)
    at process.internalTickCallback (internal/process/next_tick.js:72:19)
Emitted 'error' event at:
    at emitErrorNT (net.js:1322:8)
    at process.internalTickCallback (internal/process/next_tick.js:72:19)
```
It seems `net.connect({ ipv6Only: true })` don't work on some linux, will look into this.

And this one [sequential/test-performance](https://ci.nodejs.org/job/node-test-commit-smartos/nodes=smartos17-64/21428/testReport/junit/(root)/test/sequential_test_performance/). Currently, I'm not sure if it's related this this PR.

oyyd commented:
I have test `uv_tcp_bind` and `uv_tcp_connect` on Centos6, and it works as expected:

<details>
<summary>libuv test</summary>

```c
TEST_IMPL(tcp_bind6) {
  struct sockaddr_in addr4;
  struct sockaddr_in6 addr6;
  uv_tcp_t server4;
  uv_tcp_t server6;
  int r;

  if (!can_ipv6())
    RETURN_SKIP("IPv6 not supported");

  ASSERT(0 == uv_ip4_addr("0.0.0.0", TEST_PORT, &addr4));
  ASSERT(0 == uv_ip6_addr("::", TEST_PORT, &addr6));

  r = uv_tcp_init(uv_default_loop(), &server6);
  ASSERT(r == 0);
  r = uv_tcp_bind(&server6, (const struct sockaddr*) &addr6, UV_TCP_IPV6ONLY);
  ASSERT(r == 0);
  r = uv_listen((uv_stream_t*)&server6, 128, NULL);
  ASSERT(r == 0);
  r = uv_tcp_init(uv_default_loop(), &server4);
  ASSERT(r == 0);
  r = uv_tcp_bind(&server4, (const struct sockaddr*) &addr4, 0);
  ASSERT(r == 0);
  r = uv_listen((uv_stream_t*)&server4, 128, NULL);
  ASSERT(r == 0);

  MAKE_VALGRIND_HAPPY();
  return 0;
}

```
</details>

And I noticed that if we don't specify `localAddress` or `localPort` in `net.connect()`, `handle.bind6()` won't get called so that `UV_TCP_IPV6ONLY` won't work. 

Therefore I suspect dual-stack support without `bind6()` is platform-specific, or defined by config files like `/etc/sysconfig/network`.

I have added a description about this in the doc and modified the test. Would like to see if the test will pass on other platforms.

oyyd commented:
Hi @Trott, Can you help resume the CI? :-)

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/18446/ (eventually--CI is tied up right now, probably with a Code & Learn backlog, but that link will eventually work and the CI will be there)

oyyd commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/18545/

oyyd commented:
`parallel/test-net-connect-ipv6only` failed on AIX:

```
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: connect EHOSTUNREACH :::35876 - Local (undefined:undefined)
    at internalConnect (net.js:883:16)
    at defaultTriggerAsyncIdScope (internal/async_hooks.js:294:19)
    at defaultTriggerAsyncIdScope (net.js:974:9)
    at internalTickCallback (internal/process/next_tick.js:70:11)
    at process._tickCallback (internal/process/next_tick.js:47:5)
    at Function.Module.runMain (internal/modules/cjs/loader.js:778:11)
    at startup (internal/bootstrap/node.js:300:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:833:3)
Emitted 'error' event at:
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at internalTickCallback (internal/process/next_tick.js:72:19)
    at process._tickCallback (internal/process/next_tick.js:47:5)
    [... lines matching original stack trace ...]
    at bootstrapNodeJSCore (internal/bootstrap/node.js:833:3)
```

I have no idea why this happened though.

Trott commented:
@nodejs/platform-aix ^^^^^^

oyyd commented:
I suggest removing the support for `net.Socket.connect()` in this PR as it fails on AIX and I guess we need to resolve it in libuv. The rest parts still make sense as the support for `server` is what we mostly care about.

Trott commented:
> I suggest removing the support for net.Socket.connect() in this PR as it fails on AIX and I guess we need to resolve it in libuv. The rest parts still make sense as the support for server is what we mostly care about.

Will you be doing that and rebasing? If not, should we apply a `help wanted` label to this so someone else will do it?

oyyd commented:
I will remove the `net.Socket.connect()` and rebase again.

oyyd commented:
Resume: https://ci.nodejs.org/job/node-test-pull-request/18839/

oyyd commented:
Resume (for known issues): https://ci.nodejs.org/job/node-test-commit-arm-fanned/4499/

oyyd commented:
I have removed the support for `net.Socket.connect()` and the CI is green now.

/cc @bnoordhuis @jasnell Does this still LGTY?

jasnell commented:
Yes, still LGTM

oyyd commented:
Landed in 33a25b2. Thanks!
